artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -267,102 +322,151 @@ public void deleteAddress(AddressInfo addressInfo) throws Exception {
+         }
+         targetQueue.flushOnIntermediate(() -> {
+            recoverContext();
+            performAck(nodeID, messageID, targetQueue, ackMessageOperation, false);
+         });
+         return;
+      }
+      if (reference != null) {
+         if (logger.isTraceEnabled()) {
+            logger.trace("Post ack Server " + server + " worked well for messageID=" + messageID + " nodeID=" + nodeID);
+         }
+         try {
+            targetQueue.acknowledge(reference);
+            OperationContextImpl.getContext().executeOnCompletion(ackMessageOperation);
+         } catch (Exception e) {
+            logger.warn(e.getMessage(), e);
+         }
+      } else {
+         if (logger.isDebugEnabled()) {
+            logger.debug("Post ack Server " + server + " could not find messageID = " + messageID +
+                            " representing nodeID=" + nodeID);
          }
       }
 
    }
 
-   private void sendMessage(AMQPMessage message) throws Exception {
+   /**
+    * this method returning true means the sendMessage was successful, and the IOContext should no longer be used.
+    * as the sendMessage was successful the OperationContext of the transaction will take care of the completion.
+    * The caller of this method should give up any reference to messageCompletionAck when this method returns true.
+    * */
+   private boolean sendMessage(AMQPMessage message, ACKMessageOperation messageCompletionAck) throws Exception {
+
       if (message.getMessageID() <= 0) {
          message.setMessageID(server.getStorageManager().generateID());
       }
 
-      Long internalID = (Long) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_ID);
+      String internalMirrorID = (String)AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, BROKER_ID);
+      if (internalMirrorID == null) {
+         internalMirrorID = getRemoteMirrorId(); // not pasisng the ID means the data was generated on the remote broker
+      }
+      Long internalIDLong = (Long) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_ID);
       String internalAddress = (String) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_DESTINATION);
 
-      if (internalID != null) {
-         message.setBrokerProperty(INTERNAL_ID_EXTRA_PROPERTY, internalID);
+      long internalID = 0;
+
+      if (internalIDLong != null) {
+         internalID = internalIDLong;
+      }
+
+
+      if (logger.isTraceEnabled()) {
+         logger.trace("sendMessage on server " + server + " for message " + message +
+                         " with internalID = " + internalIDLong + " mirror id " + internalMirrorID);
+      }
+
+
+      routingContext.setDuplicateDetection(false); // we do our own duplicate detection here
+
+      byte[] duplicateIDBytes = duplicateIDGenerator.generateDuplicateID(internalMirrorID, internalID);
+
+      if (duplicateIDCache.contains(duplicateIDBytes)) {
```

Comment:
Do we need to consider the 'internalAddress' value as well if it exists? (Not clear when it does exactly)

Reply:
the next commit removed this.

Reply:
you wouldn't though.. this is about detecting a duplicate send and ignore it.